dali/test/python/test_pipeline_decorator.py

```diff
@@ -0,0 +1,109 @@
+# limitations under the License.
+
+from nvidia.dali import Pipeline, pipeline_def
+from nose.tools import nottest, raises
+import nvidia.dali.fn as fn
+from test_utils import get_dali_extra_path, compare_pipelines
+import os
+
+data_root = get_dali_extra_path()
+images_dir = os.path.join(data_root, 'db', 'single', 'jpeg')
+
+N_ITER = 7
+
+max_batch_size = 16
+num_threads = 4
+device_id = 0
+
+
+def reference_pipeline(flip_vertical, flip_horizontal, ref_batch_size=max_batch_size):
+    pipeline = Pipeline(ref_batch_size, num_threads, device_id)
+    with pipeline:
+        data, _ = fn.file_reader(file_root=images_dir)
+        img = fn.image_decoder(data, device="mixed")
+        flipped = fn.flip(img, horizontal=flip_horizontal, vertical=flip_vertical)
+        pipeline.set_outputs(flipped, img)
+    return pipeline
+
+
+@nottest
+@pipeline_def(batch_size=max_batch_size, num_threads=num_threads, device_id=device_id)
+def pipeline_static(flip_vertical, flip_horizontal):
+    data, _ = fn.file_reader(file_root=images_dir)
+    img = fn.image_decoder(data, device="mixed")
+    flipped = fn.flip(img, horizontal=flip_horizontal, vertical=flip_vertical)
+    return flipped, img
+
+
+@nottest
+@pipeline_def
+def pipeline_runtime(flip_vertical, flip_horizontal):
+    data, _ = fn.file_reader(file_root=images_dir)
+    img = fn.image_decoder(data, device="mixed")
+    flipped = fn.flip(img, horizontal=flip_horizontal, vertical=flip_vertical)
+    return flipped, img
+
+
+
+
+
+@nottest
+def test_pipeline_static(flip_vertical, flip_horizontal):
+    put_args = pipeline_static(flip_vertical, flip_horizontal)
+    ref = reference_pipeline(flip_vertical, flip_horizontal)
+    compare_pipelines(put_args, ref, batch_size=max_batch_size, N_iterations=N_ITER)
+
+
+@nottest
+def test_pipeline_runtime(flip_vertical, flip_horizontal):
+    put_combined = pipeline_runtime(flip_vertical, flip_horizontal, batch_size=max_batch_size,
+                                    num_threads=num_threads, device_id=device_id)
+    ref = reference_pipeline(flip_vertical, flip_horizontal)
+    compare_pipelines(put_combined, ref, batch_size=max_batch_size, N_iterations=N_ITER)
+
+
+@nottest
+def test_pipeline_override(flip_vertical, flip_horizontal, batch_size):
+    put_combined = pipeline_static(flip_vertical, flip_horizontal, batch_size=batch_size,
+                                   num_threads=num_threads, device_id=device_id)
+    ref = reference_pipeline(flip_vertical, flip_horizontal, ref_batch_size=batch_size)
+    compare_pipelines(put_combined, ref, batch_size=batch_size, N_iterations=N_ITER)
+
+
+def test_pipeline_decorator():
+    for vert in [0, 1]:
+        for hori in [0, 1]:
+            yield test_pipeline_static, vert, hori
+            yield test_pipeline_runtime, vert, hori
+            yield test_pipeline_override, vert, hori, 16
+    yield test_pipeline_runtime, fn.random.coin_flip(seed=123), fn.random.coin_flip(seed=234)
```

Comment:
does this work? can we pass data nodes from outside `with pipe:` ... ?

Reply:
Apparently yes...

Reply:
We can if they don't have side-effects.